include/boost/compute/random/threefry.hpp

```diff
@@ -0,0 +1,246 @@
+// Copyright 2010-2012, D. E. Shaw Research.
+// All rights reserved.
+
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+
+// * Redistributions of source code must retain the above copyright
+//   notice, this list of conditions, and the following disclaimer.
+
+// * Redistributions in binary form must reproduce the above copyright
+//   notice, this list of conditions, and the following disclaimer in the
+//   documentation and/or other materials provided with the distribution.
+
+// * Neither the name of D. E. Shaw Research nor the names of its
+//   contributors may be used to endorse or promote products derived from
+//   this software without specific prior written permission.
+
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+#ifndef BOOST_COMPUTE_RANDOM_THREEFRY_HPP
+#define BOOST_COMPUTE_RANDOM_THREEFRY_HPP
+
+#include <boost/compute/types.hpp>
+#include <boost/compute/buffer.hpp>
+#include <boost/compute/kernel.hpp>
+#include <boost/compute/context.hpp>
+#include <boost/compute/program.hpp>
+#include <boost/compute/command_queue.hpp>
+#include <boost/compute/algorithm/transform.hpp>
+#include <boost/compute/detail/iterator_range_size.hpp>
+#include <boost/compute/utility/program_cache.hpp>
+#include <boost/compute/iterator/discard_iterator.hpp>
+
+namespace boost {
+namespace compute {
+namespace random {
+
+
+/// \class threefry_rng_engine
+/// \brief Threefry pseudorandom number generator.
+template<class T>
+class threefry_engine
+{
+public:
+    typedef T result_type;
+
+    /// Creates a new threefry_engine and seeds it with \p value.
+    explicit threefry_engine(command_queue &queue)
+        : m_context(queue.get_context())
+    {
+        // setup program
+        load_program();
+    }
+
+    /// Creates a new threefry_engine object as a copy of \p other.
+    threefry_engine(const threefry_engine<T> &other)
+        : m_context(other.m_context)
+    {
+    }
+
+    /// Copies \p other to \c *this.
+    threefry_engine<T>& operator=(const threefry_engine<T> &other)
+    {
+        if(this != &other){
+            m_context = other.m_context;
```

Comment:
Same as with the copy-constructor, we need to copy over the program object also.